### PR TITLE
Deal with invalid UTF-8 from `gai_strerror`

### LINKED ISSUE #127563: Panic in std
version: nightly-2023-11-22
os: centos-7

panic:
thread '' panicked at library/std/src/sys/unix/net.rs:50:76:
called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 4, error_len: Some(1) }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '' panicked at library/std/src/sys/unix/net.rs:50:76:
called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 4, error_len: Some(1) }
thread '' panicked at library/std/src/sys/unix/net.rs:50:76:
called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 4, error_len: Some(1) }
thread '' panicked at library/std/src/sys/unix/net.rs:50:76:
called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 4, error_len: Some(1) }

### PR DESCRIPTION
**Pull request by Noratrieb**
When the system is using a non-UTF-8 locale, the value will indeed not be UTF-8. That sucks for everyone involved, but is no reason for panic. We can "handle" this gracefully by just using from lossy, replacing the invalid UTF-8 with � and keeping the accidentally valid UTF-8. Good luck when debugging, but at least it's not a crash.

We already do this for `strerror_r`.

fixes #127563

### CODE REVIEW

**Comment by Nilstrieb** [comment]
mara just removed herself from rotation
r? libs

File: library/std/src/sys/pal/unix/net.rs [hunk 1]
```diff
@@ -47,7 +46,9 @@ pub fn cvt_gai(err: c_int) -> io::Result<()> {
 
     #[cfg(not(target_os = "espidf"))]
     let detail = unsafe {
-        str::from_utf8(CStr::from_ptr(libc::gai_strerror(err)).to_bytes()).unwrap().to_owned()
+        // We can't always expect a UTF-8 environment. When we don't get that luxury,
+        // it's better to give a low-quality error message than none at all.
+        String::from_utf8_lossy(CStr::from_ptr(libc::gai_strerror(err)).to_bytes())
```
Comment (eggyal) [inline library/std/src/sys/pal/unix/net.rs:51]: Perhaps use [`CStr::to_string_lossy`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_string_lossy) here instead, which will avoid allocating a new `String` if not necessary?

**Review by eggyal — CHANGES_REQUESTED** [review]
One small suggestion from me.

File: library/std/src/sys/pal/unix/net.rs (hunk 1 above)
Comment (zachs18) [reply library/std/src/sys/pal/unix/net.rs:51]: [`String::from_ut8_lossy`](https://doc.rust-lang.org/nightly/std/string/struct.String.html#method.from_utf8_lossy) also avoids allocating a new  `String` if necessary; it returns a `Cow<'_, str>` to handle that. It looks like [`CStr::to_string_lossy`](https://doc.rust-lang.org/src/alloc/ffi/c_str.rs.html#1100) just delegates to `String::from_ut8_lossy` anyway.

Comment (eggyal) [reply library/std/src/sys/pal/unix/net.rs:51]: Oh, quite right I hadn't realised that!  Ignore me 😳

**Comment by eggyal** [comment]
Also wonder whether it wouldn't be better to discard the returned string entirely and only use the static message prefix?

**Review by joboet — APPROVED** [review]
r=me with the nit applied

**Comment by workingjubilee** [comment]
> Also wonder whether it wouldn't be better to discard the returned string entirely and only use the static message prefix?

I don't really think so. We can't predict the exact encoding or why it's not valid UTF-8. In many cases, it may be partially UTF-8 (often, identical in the ASCII plane), and so if the error includes some UTF-8 data (like the classic Arabic numerals, for instance) then it's useful to continue emitting that data. In cases where it isn't, it may be that the resulting emitted text will be nonsensical in the target locale either way, even if we use the static prefix, so we shouldn't worry too much.

**Pull request merged**